SignalServiceKit/src/Profiles/OWSProfilesManager.m

```diff
@@ -105,14 +181,261 @@ - (void)observeNotifications
 + (NSData *)generateLocalProfileKey
 {
     // TODO:
-    OWSFail(@"Profile key generation is not yet implemented.");
+    DDLogVerbose(@"%@ Profile key generation is not yet implemented.", self.tag);
     return [SecurityUtils generateRandomBytes:kProfileKeyLength];
 }
 
-- (nullable NSData *)localProfileKey
+#pragma mark - Local Profile
+
+// This method is use to update client "local profile" state.
+- (void)updateLocalProfileName:(nullable NSString *)localProfileName
+       localProfileAvatarImage:(nullable UIImage *)localProfileAvatarImage
+    localProfileAvatarMetadata:(nullable AvatarMetadata *)localProfileAvatarMetadata
 {
-    OWSAssert(_localProfileKey.length == kProfileKeyLength);
-    return _localProfileKey;
+    OWSAssert([NSThread isMainThread]);
+
+    // The avatar image and filename should both be set, or neither should be set.
+    if (!localProfileAvatarMetadata && localProfileAvatarImage) {
+        OWSFail(@"Missing avatar metadata.");
+        localProfileAvatarImage = nil;
+    }
+    if (localProfileAvatarMetadata && !localProfileAvatarImage) {
+        OWSFail(@"Missing avatar image.");
+        localProfileAvatarMetadata = nil;
+    }
+
+    self.localProfileName = localProfileName;
+    self.localProfileAvatarImage = localProfileAvatarImage;
+    self.localProfileAvatarMetadata = localProfileAvatarMetadata;
+
+    if (localProfileName) {
+        [self.storageManager setObject:localProfileName
+                                forKey:kOWSProfilesManager_LocalProfileNameKey
+                          inCollection:kOWSProfilesManager_Collection];
+    } else {
+        [self.storageManager removeObjectForKey:kOWSProfilesManager_LocalProfileNameKey
+                                   inCollection:kOWSProfilesManager_Collection];
+    }
+    if (localProfileAvatarMetadata) {
+        [self.storageManager setObject:localProfileAvatarMetadata
+                                forKey:kOWSProfilesManager_LocalProfileAvatarMetadataKey
+                          inCollection:kOWSProfilesManager_Collection];
+    } else {
+        [self.storageManager removeObjectForKey:kOWSProfilesManager_LocalProfileAvatarMetadataKey
+                                   inCollection:kOWSProfilesManager_Collection];
+    }
+
+    [[NSNotificationCenter defaultCenter] postNotificationName:kNSNotificationName_LocalProfileDidChange
+                                                        object:nil
+                                                      userInfo:nil];
+}
+
+- (void)updateLocalProfileName:(nullable NSString *)localProfileName
```

Comment:
I was trying to use the name "local" in this class to distinguish profile state related to the local user and of other users.  I agree it's not ideal.  Can you think of a better term?